bundles/org.openhab.binding.freeboxos/src/main/java/org/openhab/binding/freeboxos/internal/handler/FreeDeviceHandler.java

```diff
@@ -0,0 +1,145 @@
+import org.openhab.core.library.unit.Units;
+import org.openhab.core.thing.Channel;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.binding.builder.ChannelBuilder;
+import org.openhab.core.thing.type.ChannelTypeUID;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ *
+ * @author Gaël L'hopital - Initial contribution
+ */
+@NonNullByDefault
+public abstract class FreeDeviceHandler extends HostHandler {
+    private final Logger logger = LoggerFactory.getLogger(FreeDeviceHandler.class);
+    private long uptime = -1;
+
+    public FreeDeviceHandler(Thing thing, ZoneId zoneId) {
+        super(thing, zoneId);
+    }
+
+    @Override
+    public void initialize() {
+        super.initialize();
+        try {
+            while (!checkBridgeHandler()) {
+                Thread.sleep(1000);
+            }
+            initializeChannels();
+        } catch (InterruptedException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+        }
+    }
+
+    @Override
+    protected void internalPoll() throws FreeboxException {
+        super.internalPoll();
+        fetchSystemConfig();
+    }
+
+    protected abstract DeviceConfig getDeviceConfig() throws FreeboxException;
+
+    protected abstract void internalCallReboot() throws FreeboxException;
+
+    private void initializeChannels() {
+        List<Channel> channels = new ArrayList<>(getThing().getChannels());
+
+        try {
+            DeviceConfig systemConfig = getDeviceConfig();
+            List<Sensor> sensors = systemConfig.getAllSensors();
+            sensors.forEach(sensor -> {
+                ChannelUID sensorId = new ChannelUID(thing.getUID(), PROPERTY_SENSORS, sensor.getId());
+                if (channels.stream().noneMatch(c -> c.getUID().equals(sensorId))) {
+                    ChannelBuilder channelBuilder = ChannelBuilder.create(sensorId).withLabel(sensor.getName());
+                    if (sensor.getKind() == SensorKind.FAN) {
+                        channels.add(channelBuilder.withAcceptedItemType(CoreItemFactory.NUMBER)
+                                .withType(new ChannelTypeUID(BINDING_ID + ":fanspeed")).build());
+                    } else if (sensor.getKind() == SensorKind.TEMP) {
+                        channels.add(channelBuilder.withAcceptedItemType("Number:Temperature")
+                                .withType(new ChannelTypeUID(BINDING_ID + ":temperature")).build());
+                    }
+                }
+            });
+            updateThing(editThing().withChannels(channels).build());
+        } catch (FreeboxException e) {
+            logger.warn("Error getting list of optional channels : {}", e.getMessage());
+        }
+    }
+
+    private void fetchSystemConfig() throws FreeboxException {
```

Comment:
Please update the thing properties (hardware version, serial number, ...).

Reply:
I did not do this because from my stand point a serial number or hardware version is not really likely to change. Your thoughts ?

Reply:
Because they are currently only set by the discovery code. On a thing created manually, they will not be set. 
In fact, you should probably move the code setting these properties from the discovery stuff to the thing handler.

Reply:
Yes, obviously this is no good. I'll have to change this so all properties are populated at first activation of the handler instead of discovery.

Reply:
Yes, exactly